ambry-store/src/main/java/com.github.ambry.store/DefaultCompactionPolicy.java

```diff
@@ -0,0 +1,144 @@
+    this.time = time;
+    this.messageRetentionTimeInMs = storeConfig.storeDeletedMessageRetentionDays * Time.SecsPerDay * Time.MsPerSec;
+  }
+
+  @Override
+  public CompactionDetails getCompactionDetails(long totalCapacity, long usedCapacity, long segmentCapacity,
+      long segmentHeaderSize, List<String> logSegmentsNotInJournal, BlobStoreStats blobStoreStats)
+      throws StoreException {
+    CompactionDetails details = null;
+    if (usedCapacity >= (storeConfig.storeMinUsedCapacityToTriggerCompactionInPercentage / 100.0) * totalCapacity) {
+      if (logSegmentsNotInJournal != null) {
+        Pair<Long, NavigableMap<String, Long>> validDataPerLogSegments = blobStoreStats.getValidDataSizeByLogSegment(
+            new TimeRange(time.milliseconds() - messageRetentionTimeInMs, 0));
+
+        Iterator<Map.Entry<String, Long>> validEntryIterator =
+            validDataPerLogSegments.getSecond().entrySet().iterator();
+        while (validEntryIterator.hasNext()) {
+          Map.Entry<String, Long> validEntry = validEntryIterator.next();
+          if (!logSegmentsNotInJournal.contains(validEntry.getKey())) {
+            validEntryIterator.remove();
+          }
+        }
+
+        CostBenefitInfo bestCandidateToCompact =
+            getBestCandidateToCompact(validDataPerLogSegments.getSecond(), segmentCapacity, segmentHeaderSize,
+                blobStoreStats.getMaxBlobSize());
+        if (bestCandidateToCompact != null) {
+          List<String> segmentsToCompact = new ArrayList<>(validDataPerLogSegments.getSecond()
+              .subMap(bestCandidateToCompact.getFirstLogSegmentName(), true,
+                  bestCandidateToCompact.getLastLogSegmentName(), true)
+              .keySet());
+          details = new CompactionDetails(validDataPerLogSegments.getFirst(), segmentsToCompact);
+        }
+      }
+    }
+    return details;
+  }
+
+  /**
+   * Finds the best candidate to compact by finding the candidate with the best cost benefit ratio
+   * @param validDataPerLogSegments the valid data size for each log segment in the form of a {@link NavigableMap} of segment names to
+   * valid data sizes.
+   * @param segmentCapacity Segment capacity of one {@link LogSegment}
+   * @param segmentHeaderSize Segment header size of a {@link LogSegment}
+   * @param maxBlobSize max blob size to factor in when calculating the cost benefit ratio
+   * @return the {@link CostBenefitInfo} for the best candidate to compact. {@code null} if there isn't any.
+   */
+  private CostBenefitInfo getBestCandidateToCompact(NavigableMap<String, Long> validDataPerLogSegments,
+      long segmentCapacity, long segmentHeaderSize, long maxBlobSize) {
+    Map.Entry<String, Long> firstEntry = validDataPerLogSegments.firstEntry();
+    Map.Entry<String, Long> lastEntry = validDataPerLogSegments.lastEntry();
+    CostBenefitInfo bestCandidateToCompact = null;
+    while (firstEntry != null && firstEntry.getKey().compareTo(lastEntry.getKey()) <= 0) {
```

Comment:
not sure why the second condition is required - `lastEntry` never changes right?